core/iwasm/common/gc/gc_common.c

```diff
@@ -165,31 +214,107 @@ wasm_struct_obj_new_with_type(WASMExecEnv *exec_env, WASMStructType *type)
 }
 
 WASMArrayObjectRef
-wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx)
-{
-    /* TODO */
+wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx,
+                                uint32 length, wasm_value_t *init_value)
+{
+    WASMArrayObjectRef array_obj;
+    WASMRttTypeRef rtt_type;
+    WASMModuleCommon *module = wasm_exec_env_get_module(exec_env);
+    WASMModuleInstanceCommon *module_inst =
+        wasm_exec_env_get_module_inst(exec_env);
+    WASMModule *wasm_module = (WASMModule *)module;
+    WASMType *defined_type = wasm_get_defined_type(module, type_idx);
```

Comment:
The `module` field in `module_inst` can be `WASMModule` if `module_inst` is instance of interpreter mode, or `AOTModule` if `module_inst` is instance of aot mode, should handle it respectively, refer to the handle of `wasm_struct_obj_new_with_typeidx` in:
https://github.com/bytecodealliance/wasm-micro-runtime/pull/2154